{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f19cfc63",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique rows only in one file: 510\n",
      "    reference  gene parent aa parent codon  aa position mutated aa  \\\n",
      "1      1_alaA  alaA       LEU          CTT        107.0        PHE   \n",
      "2      2_alaA  alaA       LEU          CTT        107.0        PHE   \n",
      "3      3_alaA  alaA       LEU          CTT        107.0        PHE   \n",
      "4      4_alaA  alaA       LEU          CTT        107.0        PHE   \n",
      "5      5_alaA  alaA       LEU          CTT        107.0        PHE   \n",
      "..        ...   ...       ...          ...          ...        ...   \n",
      "258  258_alaA  alaA       ASP          GAT         92.0        HIS   \n",
      "259  259_alaA  alaA       ASP          GAT         92.0        HIS   \n",
      "260  260_alaA  alaA       ASP          GAT         92.0        HIS   \n",
      "261  261_alaA  alaA       ASP          GAT         92.0        HIS   \n",
      "262  262_alaA  alaA       ASP          GAT         92.0        HIS   \n",
      "\n",
      "    child codon  nt position  dist mut pam  pam mutated pam  \\\n",
      "1           TTC          319           -18  CGG           -   \n",
      "2           TTT          319           -12  TGG           -   \n",
      "3           TTC          319           -12  TGG           -   \n",
      "4           TTT          319            -4  CGG         CAG   \n",
      "5           TTC          319            -4  CGG         CAG   \n",
      "..          ...          ...           ...  ...         ...   \n",
      "258         CAC          274           -16  AGG         AAG   \n",
      "259         CAT          274            -9  TGG           -   \n",
      "260         CAC          274            -9  TGG           -   \n",
      "261         CAT          274           -18  CCA           -   \n",
      "262         CAC          274           -18  CCA           -   \n",
      "\n",
      "                                          homology arm  \\\n",
      "1    atgcgtgatgttaccgtggaagatatttacatcggcaatggtgtat...   \n",
      "2    cgtgatgttaccgtggaagatatttacatcggcaatggtgtatcgg...   \n",
      "3    cgtgatgttaccgtggaagatatttacatcggcaatggtgtatcgg...   \n",
      "4    atgttaccgtggaagatatttacatcggcaatggtgtaTCAgagTT...   \n",
      "5    atgttaccgtggaagatatttacatcggcaatggtgtaTCAgagTT...   \n",
      "..                                                 ...   \n",
      "258  actccgcgcgtaaagccatcatgcagcactacCAAgctcgtggcat...   \n",
      "259  ccgcgcgtaaagccatcatgcagcactaccaggctcgtggcatgcg...   \n",
      "260  ccgcgcgtaaagccatcatgcagcactaccaggctcgtggcatgcg...   \n",
      "261  tactccgcgcgtaaagccatcatgcagcactaccaggctcgtggca...   \n",
      "262  tactccgcgcgtaaagccatcatgcagcactaccaggctcgtggca...   \n",
      "\n",
      "    target strand of base pairing region   base pairing region  \\\n",
      "1           template / non coding strand  CCGTGGAAGATATTTACATC   \n",
      "2           template / non coding strand  AAGATATTTACATCGGCAAT   \n",
      "3           template / non coding strand  AAGATATTTACATCGGCAAT   \n",
      "4           template / non coding strand  TTACATCGGCAATGGTGTAT   \n",
      "5           template / non coding strand  TTACATCGGCAATGGTGTAT   \n",
      "..                                   ...                   ...   \n",
      "258         template / non coding strand  AGCCATCATGCAGCACTACC   \n",
      "259         template / non coding strand  TGCAGCACTACCAGGCTCGT   \n",
      "260         template / non coding strand  TGCAGCACTACCAGGCTCGT   \n",
      "261         non template / coding strand  CATCACGCATGCCACGAGCC   \n",
      "262         non template / coding strand  CATCACGCATGCCACGAGCC   \n",
      "\n",
      "                                                 oligo  \n",
      "1    TCCTCTGGCGGAAAGCCTatgcgtgatgttaccgtggaagatattt...  \n",
      "2    TCCTCTGGCGGAAAGCCTcgtgatgttaccgtggaagatatttaca...  \n",
      "3    TCCTCTGGCGGAAAGCCTcgtgatgttaccgtggaagatatttaca...  \n",
      "4    TCCTCTGGCGGAAAGCCTatgttaccgtggaagatatttacatcgg...  \n",
      "5    TCCTCTGGCGGAAAGCCTatgttaccgtggaagatatttacatcgg...  \n",
      "..                                                 ...  \n",
      "258  TCCTCTGGCGGAAAGCCTactccgcgcgtaaagccatcatgcagca...  \n",
      "259  TCCTCTGGCGGAAAGCCTccgcgcgtaaagccatcatgcagcacta...  \n",
      "260  TCCTCTGGCGGAAAGCCTccgcgcgtaaagccatcatgcagcacta...  \n",
      "261  TCCTCTGGCGGAAAGCCTtactccgcgcgtaaagccatcatgcagc...  \n",
      "262  TCCTCTGGCGGAAAGCCTtactccgcgcgtaaagccatcatgcagc...  \n",
      "\n",
      "[510 rows x 15 columns]\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load both CSV files\n",
    "df1 = pd.read_csv('../Example_Data/sgRNA_insert_list.csv')\n",
    "df2 = pd.read_csv('../Example_Data/example_sgRNA_inserts.csv')\n",
    "\n",
    "\n",
    "# Concatenate and find symmetric difference\n",
    "combined = pd.concat([df1, df2]).drop_duplicates(keep=False)\n",
    "\n",
    "# Show result\n",
    "print(f\"Number of unique rows only in one file: {len(combined)}\")\n",
    "print(combined)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e5e1c73",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ensure the columns are in the same order and have the same names\n",
    "df1 = df1.sort_index(axis=1)\n",
    "df2 = df2.sort_index(axis=1)\n",
    "\n",
    "# Merge with indicator to find which rows are unique to each\n",
    "merged = pd.merge(df1, df2, how='outer', indicator=True)\n",
    "\n",
    "# Filter rows that are only in one of the two dataframes\n",
    "diff_rows = merged[merged['_merge'] != 'both'].drop(columns=['_merge'])\n",
    "\n",
    "print(f\"Number of rows only in one file: {len(diff_rows)}\")\n",
    "print(diff_rows)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
