{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f19cfc63",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique rows only in one file: 516\n",
      "     Unnamed: 0 reference  gene parent aa parent codon  aa position  \\\n",
      "0           0.0    0_alaA  alaA       LEU          CTT        107.0   \n",
      "1           1.0    1_alaA  alaA       LEU          CTT        107.0   \n",
      "2           2.0    2_alaA  alaA       LEU          CTT        107.0   \n",
      "3           3.0    3_alaA  alaA       LEU          CTT        107.0   \n",
      "4           4.0    4_alaA  alaA       LEU          CTT        107.0   \n",
      "..          ...       ...   ...       ...          ...          ...   \n",
      "258         NaN  258_alaA  alaA       ASP          GAT         92.0   \n",
      "259         NaN  259_alaA  alaA       ASP          GAT         92.0   \n",
      "260         NaN  260_alaA  alaA       ASP          GAT         92.0   \n",
      "261         NaN  261_alaA  alaA       ASP          GAT         92.0   \n",
      "262         NaN  262_alaA  alaA       ASP          GAT         92.0   \n",
      "\n",
      "    mutated aa child codon  nt position  dist mut pam  pam mutated pam  \\\n",
      "0          PHE         TTT          319           -18  CGG           -   \n",
      "1          PHE         TTC          319           -18  CGG           -   \n",
      "2          PHE         TTT          319           -12  TGG           -   \n",
      "3          PHE         TTC          319           -12  TGG           -   \n",
      "4          PHE         TTT          319            -4  CGG         CAG   \n",
      "..         ...         ...          ...           ...  ...         ...   \n",
      "258        HIS         CAC          274           -16  AGG         AAG   \n",
      "259        HIS         CAT          274            -9  TGG           -   \n",
      "260        HIS         CAC          274            -9  TGG           -   \n",
      "261        HIS         CAT          274           -18  CCA           -   \n",
      "262        HIS         CAC          274           -18  CCA           -   \n",
      "\n",
      "                                          homology arm  \\\n",
      "0    atgcgtgatgttaccgtggaagatatttacatcggcaatggtgtat...   \n",
      "1    atgcgtgatgttaccgtggaagatatttacatcggcaatggtgtat...   \n",
      "2    cgtgatgttaccgtggaagatatttacatcggcaatggtgtatcgg...   \n",
      "3    cgtgatgttaccgtggaagatatttacatcggcaatggtgtatcgg...   \n",
      "4    atgttaccgtggaagatatttacatcggcaatggtgtaTCAgagTT...   \n",
      "..                                                 ...   \n",
      "258  actccgcgcgtaaagccatcatgcagcactacCAAgctcgtggcat...   \n",
      "259  ccgcgcgtaaagccatcatgcagcactaccaggctcgtggcatgcg...   \n",
      "260  ccgcgcgtaaagccatcatgcagcactaccaggctcgtggcatgcg...   \n",
      "261  tactccgcgcgtaaagccatcatgcagcactaccaggctcgtggca...   \n",
      "262  tactccgcgcgtaaagccatcatgcagcactaccaggctcgtggca...   \n",
      "\n",
      "    target strand of base pairing region   base pairing region  \\\n",
      "0           template / non coding strand  CCGTGGAAGATATTTACATC   \n",
      "1           template / non coding strand  CCGTGGAAGATATTTACATC   \n",
      "2           template / non coding strand  AAGATATTTACATCGGCAAT   \n",
      "3           template / non coding strand  AAGATATTTACATCGGCAAT   \n",
      "4           template / non coding strand  TTACATCGGCAATGGTGTAT   \n",
      "..                                   ...                   ...   \n",
      "258         template / non coding strand  AGCCATCATGCAGCACTACC   \n",
      "259         template / non coding strand  TGCAGCACTACCAGGCTCGT   \n",
      "260         template / non coding strand  TGCAGCACTACCAGGCTCGT   \n",
      "261         non template / coding strand  CATCACGCATGCCACGAGCC   \n",
      "262         non template / coding strand  CATCACGCATGCCACGAGCC   \n",
      "\n",
      "                                                 oligo  \n",
      "0    TCCTCTGGCGGAAAGCCTatgcgtgatgttaccgtggaagatattt...  \n",
      "1    TCCTCTGGCGGAAAGCCTatgcgtgatgttaccgtggaagatattt...  \n",
      "2    TCCTCTGGCGGAAAGCCTcgtgatgttaccgtggaagatatttaca...  \n",
      "3    TCCTCTGGCGGAAAGCCTcgtgatgttaccgtggaagatatttaca...  \n",
      "4    TCCTCTGGCGGAAAGCCTatgttaccgtggaagatatttacatcgg...  \n",
      "..                                                 ...  \n",
      "258  TCCTCTGGCGGAAAGCCTactccgcgcgtaaagccatcatgcagca...  \n",
      "259  TCCTCTGGCGGAAAGCCTccgcgcgtaaagccatcatgcagcacta...  \n",
      "260  TCCTCTGGCGGAAAGCCTccgcgcgtaaagccatcatgcagcacta...  \n",
      "261  TCCTCTGGCGGAAAGCCTtactccgcgcgtaaagccatcatgcagc...  \n",
      "262  TCCTCTGGCGGAAAGCCTtactccgcgcgtaaagccatcatgcagc...  \n",
      "\n",
      "[516 rows x 16 columns]\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load both CSV files\n",
    "df1 = pd.read_csv('../Example_Data/sgRNA_insert_list.csv')\n",
    "df2 = pd.read_csv('../Example_Data/example_sgRNA_inserts.csv')\n",
    "\n",
    "\n",
    "# Concatenate and find symmetric difference\n",
    "combined = pd.concat([df1, df2]).drop_duplicates(keep=False)\n",
    "\n",
    "# Show result\n",
    "print(f\"Number of unique rows only in one file: {len(combined)}\")\n",
    "print(combined)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
